KEP-4815 DRA Partitionable devices design update (#5066)
kubernetes/enhancements: keps/sig-scheduling/4815-dra-partitionable-devices/README.md

  devices and fixed-size partitions of full devices (across multiple
  dimensions). Both full devices and their valid set of partitions must be
  explicitly listed in a ResourceSlice for the scheduler to consider them for
  allocation.

* Keep the user-facing mechanism to request a full device or one of its
  partitions the same. In 1.31, it is already possible to select a device
  partition via a `Resourceclaim`, so long as the device vendor has
  pre-partitioned its devices and advertised them in a `ResourceSlice`. The
  user-facing mechanism to select these devices shouldn't change -- only the
  ability for the vendor to advertise "overlapping" partitions, such that the
  scheduler will never allocate conflicting ones at the same time.

* Provide abstractions to keep the overall footprint of a `ResourceSlice`
  small, such that all full devices (and their partitions) can be represented
  in a concise way.

### Non-Goals

* Allow a user to allocate arbitrarily-sized partitions of a device. In other
  words, only those partitions explicitly defined by a `ResourceSlice` will be
  allocatable by a user. Arbitrary slicing is out of scope for this KEP.

## Proposal

> On a high level, this proposal is based around two new concepts:

Reviewer: ```suggestion
At a high level, this proposal is based around two new concepts:
```

1. Capacity pool, which is a construct for defining the capacities available for

>    sharing among partitions of a single physical device. Drivers/controllers will

Reviewer: ```suggestion
   sharing among partitions of a single physical device. Drivers will
```

The implementation of these concepts requires several changes to the
ResourceSlice API. It introduces two new fields on the `ResourceSliceSpec`.
`CapacityPools` defines a list of `CapacityPools` while `Mixins` define the
mixins.

1. The `CapacityPools` field is a list of named `CapacityPool`s. Each
   defines a set of capacities that is available for devices. This makes it possible
   to define overlapping partitions of devices, while still making sure that no
   device can be allocated if the necessary capacity is not available. A `CapacityPool`
   can optionally reference one or more `CapacityPoolMixin`s using the `Includes` field
   to extend the set of capacities it defines explicitly.

1. The `Mixins` field is of type `ResourceSliceMixin`, which has fields for the three
   types of mixins available in a ResourceSlice. These are `CapacityPool`,
   `Device`, and `DeviceCapacityConsumption`.

    1. The `CapacityPool` field defines a list of named `CapacityPoolMixin`s. These
      define a set of capacities that can be used to extend the capacities explicitly
      defined in a `CapacityPool`. This allows for reduced duplication if there are many

>       identical physical devices that must be represented as capacity pools. CapacityPoolMixins

Reviewer: ```suggestion
      identical physical devices that must be represented as capacity pools. `CapacityPoolMixin`s
```

      can not be referenced directly by devices.

    1. The `Device` field is a list of named `DeviceMixin`s. These define
      attributes and capacities that can be used to extend what is defined

>       explicitly in a `CompositeDevice`. Mixins can not be allocated directly,

Reviewer: This is the first time that `CompositeDevice` is mentioned in the revised version of this KEP.#

As such, let's reword this to the following:
```suggestion
      explicitly in a new device type called `CompositeDevice` (introduced in more detail below). Mixins cannot be allocated directly,
```

Reviewer 2: Side note: for now in this KEP revision we assume that we keep `CompositeDevice`. That's correct, no need to change anything.

In https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?tab=t.0 I proposed some changes that would make it safe to flatten the `Device` struct. This is something that we can try to describe in the next KEP update (https://github.com/kubernetes/enhancements/pull/5069), or treat it as an implementation choice and retro-actively update the KEP after we have agreed on the implementation in https://github.com/kubernetes/kubernetes/pull/129970. Given the time constraints I am leaning towards the latter.

- name: gpu-0-partition-1
  composite:
    includes:
    - name: gpu-partition-capacity
    consumesCapacity:
    - capacityPool: gpu-pool-mixin
    - includes:
      - name: gpu-partition-consumption
- name: gpu-0-partition-2
  composite:
    includes:
    - name: gpu-partition-capacity
    consumesCapacity:
    - capacityPool: gpu-pool-mixin
    - includes:
      - name: gpu-partition-consumption
- name: gpu-0-partition-3
  composite:
    includes:
    - name: gpu-partition-capacity
    consumesCapacity:
    - capacityPool: gpu-pool-mixin
    - includes:
      - name: gpu-partition-consumption
```

In this example, five devices are defined: a full GPU called "gpu-0" and four
partitions of "gpu-0" called "gpu-0-partion-0", "gpu-0-partion-1",
"gpu-0-partion-2", and "gpu-0-partion-3" respectively. The physical GPU is
represented as a shared capacity pool, from which all five devices draw their
capacity. It also sets up a device mixin for the partitions that defines the
advertised resource availability and a device capacity consumption mixin that
defines the capacity the device draws from the capaity pool. This allows the
definition of each partition to be quite compact.

> After all mixins have been "flattened", the scheduler end up with a list

Reviewer: ```suggestion
After all mixins have been "flattened", the scheduler ends up with a list
```

## Proposal

On a high level, this proposal is based around two new concepts:

1. Capacity pool, which is a construct for defining the capacities available for
   sharing among partitions of a single physical device. Drivers/controllers will
   typically create a capacity pool that specifies all the available capacities of
   a physical device and then define the partitions available for DRA. Each
   partition will specify how much capacity it will consume from the capacity
   pool, which allows the scheduler to keep track of available capacity in
   a pool and decide whether a partition can be allocated. A device will
   separately define the capacities that are accessible to the user for the partition.

1. Mixins, which allows devices and resource pools to reference a mixin
   object that defines some properties and that when used, will extend the properties
   they define explicitly on the device or resource pool. The objective is to reduce
   the duplication and overall footprint of `ResourceSlices`.

> 

Reviewer: ```suggestion
```

  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting capacities from those provided via other mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Capacity map[QualifiedName]DeviceCapacity
}

// DeviceCapacityConsumptionMixin defines a mixin that composite
// devices can include to adopt the consuption capacity defined in
// the mixin.
type DeviceCapacityConsumptionMixin struct {
  // Name is a unique identifier among all device capacity consumption 
  // mixins in the ResourceSlice. It must be a DNS label.
  //
  // +required
  Name string

  // CapacityConsumption defines a set of capacities
  // that a device will consume from a capacity pool.
  //
  // The capacity pool is not specified here but is determined
  // from the context in which the DeviceCapacityConsumptionMixin
  // is referenced from the device.
  // 
  // The maximum number of capacities is 32
  //
  // +required

>   CapacityConsumption map[QualifiedName]DeviceCapacity

Reviewer: I think we can just call this `capacity` -- its clear from where its used, what the context is.

Author: Yeah, agree. I think the name of this field should be the same in both the mixin and the device (i.e. the `DeviceCapacityConsumption` struct) should match, so I've updated both.

  //
  // +optional
  // +listType=atomic
  Includes []DeviceMixinRef

  // Attributes defines the set of attributes for this device.
  // The name of each attribute must be unique in that set.
  //
  // To ensure this uniqueness, attributes defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting attributes from those provided via mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Attributes map[QualifiedName]DeviceAttribute

  // Capacity defines the set of capacities for this mixin.
  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting capacities from those provided via other mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional

>   Capacity map[QualifiedName]DeviceCapacity

Reviewer: Does it make sense to rename this `AdvertisedCapacity` to contrast it with `ConsumedCapacity`?

Author: Yeah, I did think about a more descriptive name here. I think it makes sense, but I am not sure if this makes the mapping from devices in the `ResourceSlice` to the properties of the CEL expressions in the `ResourceClaims` less obvious? 
I was also uncertain about introducing differences in naming between `BasicDevice` and `CompositeDevice`, although that would be something we can fix in `v1beta2`.

Reviewer 3: My $0.02 - it's fine as is.

  // ConsumesCapacity defines a list of references to capacity
  // pools and the set of capacities that the device will
  // consume from those pools.
  //
  // The capacities can be defined either by referencing one
  // or more DeviceCapacityConsumptionMixins for by listing
  // the capacities directly. The latter will always override
  // any capacities coming in from the mixins.
  //
  // The maximum number of references is 32. This is the
  // same as the maximum number of capacity pools allowed in
  // a ResourceSlice.
  //
  // +required
  // +listType=atomic

>   ConsumesCapacity []CapacityConsumptionRef

Reviewer: ```suggestion
  ConsumedCapacity []DeviceCapacityConsumptionRef
```

  // Capacity defines the set of capacities for this capacity pool
  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Capacities listed here will always take precedence over any included
  // from a mixin.
  //
  // The maximum number of capacities is 32.
  //
  // +required
  Capacity map[QualifiedName]DeviceCapacity
}

// DeviceMixinRef defines a reference to a device mixin.
type DeviceMixinRef struct {
  // Name refers to the name of a device mixin in the pool.
  //
  // +required
  Name string
}

// CapacityPoolMixinRef defines a reference from a capacity pool
// to a capacity pool mixin.
type CapacityPoolMixinRef struct {
  // Name is the name of a CapacityPoolMixin.
  //
  // +required
  Name string
}

type DeviceCapacityConsumptionMixinRef struct {
  // Name is the name of a DeviceCapacityConsumptionMixin.
  //
  // +required
  Name string
}

// DeviceCapacityConsumptionRef defines a set of capacities that
// a device will consume from a capacity pool.

> type DeviceCapacityConsumptionRef struct {

Author: Yeah, agree that it could have a better name. Since this specifies the capacity consumption for a device, I think just removing the `Ref` suffix works pretty well. So it is updated to just be `DeviceCapacityConsumption` now.

  // CapacityPool defines the capacity pool from which the
  // capacities defined (either directly or through a mixin)
  // will be consumed from.
  //
  // +required
  CapacityPool string
  
  // Mixins defines a list of references to DeviceCapacityConsumptionMixins.
  // The capacities listed in these will be included in among the
  // capacities that will be consumed by the device.
  //
  // Capacities listed directly will override any capacities coming
  // from mixins.
  //
  // +optional
  Mixins []DeviceCapacityConsumptionMixinRef

  // CapacityConsumption defines the capacity that will be consumed by
  // the device.
  //
  // Capacity listed here will override any capacities that
  // are also defined in any of the referenced mixins.
  //
  // +optional

>   CapacityConsumption DeviceCapacity

Author: Yeah, I'm not sure what I was thinking when I named the field `Mixins`... But fixed now.

}
```

As mentioned previously, the main features being added here are (1) the ability
to include a set of mixins in a device or capacity pool definition, and (2) the
ability to express that multiple devices draw from the same pool of capacity, so
allocation of one device might make other devices unallocatable.

To simplify the conversation, we discuss each new feature separately, starting
with "mixins" for both capacity pools and devices, which allows a set of mixins to
extend the properties defined explicitly.

### Extending device and capacity pool with a set of mixins

A simple example the defines a set of mixins and includes them in the
definition of 4 NVIDIA A100 GPUs can be seen below:

```yaml
capacityPoolMixins:
- name: gpu-pool-mixin
  capacity:

>     memory: 40Gi

Reviewer: I believe this should be (all throughout):
```
  capacity:
    memory:
      value: 40Gi
```

Author: I've cleaned up the updated examples and will make sure do it this way for the rest of the examples.

## Proposal

On a high level, this proposal is based around two new concepts:

1. Capacity pool, which is a construct for defining the capacities available for
   sharing among partitions of a single physical device. Drivers/controllers will
   typically create a capacity pool that specifies all the available capacities of
   a physical device and then define the partitions available for DRA. Each
   partition will specify how much capacity it will consume from the capacity
   pool, which allows the scheduler to keep track of available capacity in
   a pool and decide whether a partition can be allocated. A device will
   separately define the capacities that are accessible to the user for the partition.

1. Mixins, which allows devices and resource pools to reference a mixin
   object that defines some properties and that when used, will extend the properties
   they define explicitly on the device or resource pool. The objective is to reduce
   the duplication and overall footprint of `ResourceSlices`.

The implementation of these concepts requires several changes to the
ResourceSlice API:

> 1. It introduces four new fields on the `ResourceSliceSpec`. First, `CapacityPools`

Author: Fixed.

>    defines a list of `CapacityPools`, each of which defines a set of capacities
>    that devices can draw from. `CapacityPoolMixins`, `DeviceMixins`, and

Reviewer 2: Nit:

```suggestion
   defines a list of `CapacityPools`. `CapacityPoolMixins`, `DeviceMixins`, and
```

Details about `CapacityPools` are available below. Having one place to learn about them seems better to me.

   can optionally reference one or more `CapacityPoolMixin`s using the `Includes` field
   to extend the set of capacities it defines explicitly.

1. The `CapacityPoolMixins` field defines a list of named `CapacityPoolMixin`s. These
   define a set of capacities that can be used to extend the capacities explicitly
   defined in a `CapacityPool`. This allows for reduced duplication if there are many
   identical physical devices that must be represented as capacity pools. CapacityPoolMixins
   can not be referenced directly by devices.

1. The `DeviceMixins` field is a list of named `DeviceMixin`s. These define
   attributes and capacities that can be used to extend what is defined
   explicitly in a `CompositeDevice`. Mixins can not be allocated directly,
   but can only be referenced by composite devices.

1. Introduce a new device type called `CompositeDevice` which has the same
   fields as a `BasicDevice`, plus two more. The first is a field called
   `Includes` and lists the device mixins for the device, while the second is
   called `ConsumesCapacity` and defines the capacity the device will draw
   from the capacity pools.

> 1. The `Includes` field serves to reference a set of `DeviceMixin`s that a

Author: Agree. I've fixed this and also used a similar structure underneath the `Mixins` field added based on the suggestion in https://github.com/kubernetes/enhancements/pull/5066#discussion_r1948953578

   will draw from the capacity pool when allocated. Therefore, if this capacity isn't
   available in the pool, the device can not be allocated. Only references
   to capacity pools in the same `ResourceSlice` is supported.

With these additions in place, the scheduler has everything it needs to support
the dynamic allocation of both full devices and their (possibly overlapping)
fixed-size partitions. That is to say, the scheduler now has the ability to
"flatten" all devices and capacity pools by applying any referenced mixins as
well as track the capacities consumed by allocated devices. More details
on the actual algorithm the scheduler follows to make allocation decisions
based on the capacity pools can be found in the Design Details section below.

## Design Details

The exact set of proposed API changes can be seen below:
```go
// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
type ResourceSliceSpec struct {
  ...

  // CapacityPools defines a list of capacity pools, each of which
  // has a name and a list of capacities available in the pool.
  //
  // The names of the pools must be unique in the ResourceSlice.
  //

>   // The maximum number of pools is 32.

Reviewer 2: The limit for the other new items (mixins, composite devices) is defined so that the overall number is the same as the before for basic device, so the maximum size should be about the same as before.

This here is allowed to add something on top of that. Each pool is small, so this should be fine.

I'm mostly just wondering how we can be a bit more sure that it's really okay. Doing a back-of-the-envelope calculation as in the past is getting more difficult. Perhaps we should have a validation unit test which creates a slice with all fields at their maximum and then checks the size of protobuf encoding?

I had that idea before, but I don't remember whether I implemented it... :sweat_smile: 

There's no need to write to etcd, [`ResourceSlice.Size`](https://github.com/kubernetes/kubernetes/blob/9e555875e79f60a238952e91e91f79f32c053f9c/staging/src/k8s.io/api/resource/v1beta1/generated.pb.go#L3340) accurately predicts the size.

This could be added to the "unit tests" section of this KEP.

Author: I couldn't find any test like this in the code base, so added a short section to the "unit tests" section describing that we plan to add this.

  //
  // +optional

>   // listType=atomic

Reviewer 2: For me, the deciding factor is whether this slice has one owner or many and how it gets updated.

When using `listType=map`, the tracking information for who owns which entry explodes. This is unnecessary here because there is a single owner, the DRA driver. There's also no need for patching specific entries.

>   // listType=atomic

Reviewer 3: Right, it was the `managedFields` argument.  This is more a note for the declarative validation effort than  for this KEP.

  // that a device will consume from a capacity pool.
  //
  // This makes it possible to define a set of shared capacities that
  // are not tied to a specific pool. The pool is inferred by context
  // in which the DeviceCapacityConsumptionMixin is referenced from
  // the device.
  //
  // The total number of device mixins, device capacity consumption mixins, 
  // capacity pool mixins, basic devices, and composite devices must be
  // less than 128.
  //
  // +optional
  // +listType=atomic
  DeviceCapacityConsumptionMixins []DeviceCapacityConsumptionMixin

  // CapacityPoolMixins represents a list of capacity pool mixins, i.e.
  // a collection of capacities that a CapacityPool can "include"
  // to extend the set of capacities it already defines.
  //
  // The main purposes of these mixins is to reduce the memory footprint
  // of capacity pools since they can reference the mixins provided here rather
  // than duplicate them.
  //
  // The total number of device mixins, device capacity consumption mixins, 
  // capacity pool mixins, basic devices, and composite devices must be
  // less than 128.
  //
  // +optional
  // +listType=atomic

>   CapacityPoolMixins []CapacityPoolMixin

Reviewer 4: Yeah, I like this nesting :+1:

>   CapacityPoolMixins []CapacityPoolMixin

Reviewer 5: me too

>   CapacityPoolMixins []CapacityPoolMixin

Reviewer: What about even:
```
mixins:
  device: 
    attribute: ...
    capacity: ...
    capacityConsumption: ...
  capacityPool::
    capacity: ...
```
That way we don't need separate mixin types for `DeviceMixins`, `DeviceCapacityConsumptionMixin`, and `CapacityPoolMixin`. We just have `AttributeMixin` and `CapacityMixin` and they get used where appropriate

>   CapacityPoolMixins []CapacityPoolMixin

Reviewer 5: works for me. @mortent will be back tomorrow

>   CapacityPoolMixins []CapacityPoolMixin

Author: So I think the initial proposal here to nest all mixins under a single Mixins structs is a clear improvement.

@klueska My assumption in the second suggestion is that each of the `attribute`, `capacity`, `capacityConsumption` fields within the device struct would be lists of named mixins. So more expanded it would look something like:
```
mixins:
  device: 
    attribute:
    - name: device-attributes-mixin
      attributes:
        minor:
          int: 3
    capacity:
    - name: device-advertised-capacities-mixin
      capacity:
        memory:
          value: 40Gi
    capacityConsumption:
    - name: device-consumed-capacities-mixin
      capacity:
        memory:
          value: 40Gi
  capacityPool::
    capacity:
    - name: capacity-pool-mixin
      capacity:
        memory:
          value: 40Gi
```

If this is the suggestion, I'm less convinced about this one for a few reasons:
* We end up with nested `attribute.attributes` and `capacity.capacity` fields, which feels a bit awkward.
* Re-using a `CapacityMixin` type saves us some duplication, but we loose some flexibility to evolve the mixins for capacity pools and devices separately in the future.
* We end up with four different mixin types, so a device will need three different `Includes` fields (`IncludesAttributes`, `IncludesCapacities`, and the `Includes` field nested under `consumesCapacity`.

>   CapacityPoolMixins []CapacityPoolMixin

Reviewer: Yeah, that second suggestion looks awkward. Let's stick with the single level of nesting

}

// DeviceMixin defines a specific device mixin for each device type.
// Besides the name, exactly one field must be set.
type DeviceMixin struct {
  // Name is a unique identifier among all device mixins in the ResourceSlice.
  // It must be a DNS label.
  //
  // +required
  Name string

  // Composite defines a mixin usable by a composite device.
  //
  // +optional
  // +oneOf=deviceMixinType
  Composite *CompositeDeviceMixin
}

// CompositeDeviceMixin defines a mixin that a composite device can include.
type CompositeDeviceMixin struct {
  // Attributes defines the set of attributes for this mixin.
  // The name of each attribute must be unique in that set.
  //

>   // To ensure this uniqueness, attributes defined by the vendor

Author: Yeah, we are doing validation in similar situations today and plan to add validation for this.

  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting attributes from those provided via other mixins are
  // overwritten by the ones provided here.
  //

>   // The maximum number of attributes and capacities combined is 32.

Author: Good question. I think the primary concern we try to address is the size limit in etcd, but that doesn't really apply to the fully unpacked device (since we never write that to etcd). I'm not sure if it might impact the performance of evaluating the CEL expressions, but I wouldn't think so.

>   // The maximum number of attributes and capacities combined is 32.

Reviewer 2: I think we have more flexibility for runtime costs and thus don't need a limit for the unpacked device. If someone wants to use this feature for something complex, they have to be aware that evaluation will also become slower.

  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting attributes from those provided via mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Attributes map[QualifiedName]DeviceAttribute

  // Capacity defines the set of capacities for this mixin.
  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting capacities from those provided via other mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Capacity map[QualifiedName]DeviceCapacity

  // ConsumesCapacity defines a list of references to capacity
  // pools and the set of capacities that the device will
  // consume from those pools.
  //
  // The capacities can be defined either by referencing one

>   // or more DeviceCapacityConsumptionMixins for by listing

Reviewer 3: "...or by listing"  (says "for")

  // the capacities directly. The latter will always override
  // any capacities coming in from the mixins.
  //
  // The maximum number of device capacity consumption entries
  // is 32. This is the same as the maximum number of capacity
  // pools allowed in a ResourceSlice.
  //
  // +required
  // +listType=atomic

>   ConsumesCapacity []DeviceCapacityConsumption

Reviewer: This is just a suggestion, I forget the convention here

>   ConsumesCapacity []DeviceCapacityConsumption

Author: I don't think we have a convention here. The other fields with similar names that were present in earlier versions of the design have been renamed. I don't feel strongly about this, but I think `ConsumesCapacity` is slightly better as the field describes the capacity the device will consume once allocated.

>   ConsumesCapacity []DeviceCapacityConsumption

Reviewer: @johnbelamaric @thockin thoughts here? I remember some discussion on conventions around verbs vs nouns, past vs. present tense, etc.

>   ConsumesCapacity []DeviceCapacityConsumption

Reviewer 3: I don't recall a strong convention.  A quick survey shows the past-tense is probably more consistent, but I am open to counter-envidence.

>   ConsumesCapacity []DeviceCapacityConsumption

Reviewer: @mortent let's change this to `consumed`, as i also recall past-tense being the more common pattern here.

  //
  // The maximum number of capacities is 32.
  //
  // +required
  Capacity map[QualifiedName]DeviceCapacity
}

// CapacityPool defines a named pool of capacities
// that are available to be used by devices defined in the
// ResourceSlice.
//
// The capacities are not allocatable by themselves, but
// can be referenced by devices. When a device is allocated,
// the capacity it uses will no longer be available for use
// by other devices.
type CapacityPool struct {
  // Name defines the name of the capacity pool.
  // It must be a DNS label.
  //
  // +required
  Name string

  // Includes defines the set of capacity pool mixins that this capacity
  // pool includes.
  //
  // The propertes of each included mixin are applied to this capacity pool in
  // order. Conflicting properties from multiple mixins are taken from the
  // last mixin listed that contains them. Properties set on the capacity pool will
  // always override properties from mixins.
  //
  // The mixins referenced here must be defined in the same
  // ResourceSlice.
  //
  // The maximum number of mixins that can be included is 8.
  //
  // +optional
  // +listType=atomic

>   Includes []CapacityPoolMixinRef

Author: @klueska Is there a particular reason why the original design has references as a list of objects with just the `Name` field rather than just a list of strings? I've kept all the `<Something>Ref` types consistent with this approach, but using a separate struct is a little bit more verbose than just a list of strings.

Reviewer: It was just to leave it open to extension in the future. We do the same for the reference to rsourceclaims in a pod spec, for example.

Author: Yeah, that makes sense.

  // Capacity defines the set of capacities for this capacity pool
  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Capacities listed here will always take precedence over any included
  // from a mixin.
  //
  // The maximum number of capacities is 32.
  //
  // +required
  Capacity map[QualifiedName]DeviceCapacity
}

// DeviceMixinRef defines a reference to a device mixin.
type DeviceMixinRef struct {
  // Name refers to the name of a device mixin in the pool.
  //
  // +required

>   Name string

Author: I don't think we have any plans that would need it, but doing it this way does allow us to add it if necessary, ref https://github.com/kubernetes/enhancements/pull/5066#discussion_r1946964462. It might be that we should just use a list of strings here instead to keep it as simple as possible and then just handle it if/when we do have an actual need for any additional fields.

  Name string
}

type DeviceCapacityConsumptionMixinRef struct {
  // Name is the name of a DeviceCapacityConsumptionMixin.
  //
  // +required
  Name string
}

// DeviceCapacityConsumption defines a set of capacities that
// a device will consume from a capacity pool.
type DeviceCapacityConsumption struct {
  // CapacityPool defines the capacity pool from which the
  // capacities defined (either directly or through a mixin)
  // will be consumed from.
  //
  // +required
  CapacityPool string
  
  // Includes defines a list of references to DeviceCapacityConsumptionMixins.
  // The capacities listed in these will be included in among the
  // capacities that will be consumed by the device.
  //
  // Capacities listed directly will override any capacities coming
  // from mixins.
  //
  // The maximum number of mixins that can be included is 8.
  //
  // +optional
  Includes []DeviceCapacityConsumptionMixinRef

  // Capacity defines the capacity that will be consumed by
  // the device.
  //
  // Capacities listed here will override any capacities that
  // are also defined in any of the referenced mixins.
  //
  // The maximum number of capacities is 32.
  //
  // +optional

>   Capacity DeviceCapacity

Author: Should we also enforce that the "effective" number of capacities for a specific device (i.e. the number of capacities defined directly plus any included through mixins) also remain below the max number? I think we probably should, but just wanted to see if anyone has an opinion on this.

>   Capacity DeviceCapacity

Reviewer: I had suggested making this a `map[QualifiedName]DeviceCapacity` (which seems to have gotten lot). Is there a reason not to do this to be symmetrical with other types?

Author: Sorry, I missed that one. Yeah, it should be a map. Updated now.

## Proposal

On a high level, this proposal is based around two new concepts:

1. Capacity pool, which is a construct for defining the capacities available for
   sharing among partitions of a single physical device. Drivers/controllers will
   typically create a capacity pool that specifies all the available capacities of
   a physical device and then define the partitions available for DRA. Each
   partition will specify how much capacity it will consume from the capacity
   pool, which allows the scheduler to keep track of available capacity in
   a pool and decide whether a partition can be allocated. A device will

>    separatly set the capacities that is published for the partition.

Reviewer 5: s/separatly/separately/

s/set the capacities that is published/define the capacities that are accessible to the user/

   the duplication and overall footprint of `ResourceSlices`.

The implementation of these concepts requires several changes to the
ResourceSlice API:

1. It introduces three new fields on the `ResourceSliceSpec`. First, `SharedCapacityPools`
   defines a list of `CapacityPools`, each of which defines a set of capacities
   that devices can draw from. `CapacityPoolMixins` and `DeviceMixins` defines mixins
   for capacity pools and devices.

1. The `SharedCapacityPools` field is a list of named `CapacityPool`s. Each
   defines a set of capacities that is available for devices. This makes it possible
   to define overlapping partitions of devices, while still making sure that no
   device can be allocated if the necessary capacity is not available. A `CapacityPool`
   can optionally reference one or more `CapacityPoolMixin`s using the `Includes` field
   to extend the set of capacities it defines explicitly.

1. The `CapacityPoolMixins` field defines a list of named `CapacityPoolMixin`s. These
   define a set of capacities that can be used to extend the capacities explicitly
   defined in a `CapacityPool`. This allows for reduced duplication if there are many

>    identical physical devices that must be represented as capacity pools. Mixins

Reviewer 5: s/Mixins/CapacityPoolMixins/

>    can not be refrenced directly by devices.

Reviewer 5: s/refrenced/referenced/

1. The `DeviceMixins` field is a list of named `CompositeDeviceMixin`s. These define
   attributes, capacities and consumed capacities that can be used to extend
   what is defined explicitly in a `CompositeDevice`. The references to
   consumed capacities in the `SharedCapacityConsumed` does not include the
   pool, but instead require that the pool is specified when referenced from the 
   `CompositeDeviceMixin`. This allows the same mixin to be used across different
   capacity pools. Mixins can not be allocated directly, but can only be referenced
   by composite devices.

1. Introduce a new device type called `CompositeDevice` which has the same
   fields as a `BasicDevice`, plus two more. The first is a field called
   `Includes` and lists the device mixins for the device, while the second is
   called `SharedCapacityConsumed` and defines the capacity the device will draw
   from the capacity pools.

1. The `Includes` field serves to reference a set of `CompositeDeviceMixin`s that a
   `CompositeDevice` can reference to extend the set of attributes,

>    capacities and consumed capacity it definces explicitly. If the referenced

Reviewer 5: s/definces/defines/

  //
  // The main purposes of these mixins is to reduce the memory footprint
  // of devices since they can reference the mixins provided here rather
  // than duplicate them.
  //
  // The total number of device mixins, capacity pool mixins, basic devices,
  // and composite devices must be less than 128.
  //
  // +optional
  // +listType=atomic
  DeviceMixins []DeviceMixin

  // CapacityPoolMixins represents a list of capacity pool mixins, i.e.
  // a collection of capacities that a CapacityPool can "include"
  // to extend the set of capacities it already defines.
  //
  // The main purposes of these mixins is to reduce the memory footprint
  // of capacity pools since they can reference the mixins provided here rather
  // than duplicate them.
  //
  // The total number of device mixins, capacity pool mixins, basic devices,
  // and composite devices must be less than 128.
  //
  // +optional
  // +listType=atomic
  CapacityPoolMixins []CapacityPoolMixin

  // SharedCapacityPools defines a list of capacity pools, each of which
  // has a name and a list of capacities available in the pool.
  //
  // The names of the pools must be unique in the ResourceSlice.
  //

>   // The maximum number of pools is 32.

Reviewer 5: So in the NVIDIA case, this limits us to 32 physical GPUs in a ResourceSlice. Which is OK, since you can have more than one slice per node, if we get to very very large nodes.

Author: Yeah, it can probably be discussed if these are the right numbers, but in this case users can always spread things across multiple `ResourceSlices`. Each NVIDIA GPU (ref the A100 example) includes a relatively large number of devices, each of which has quite a few annotations and capacities. So I think there might be many reasons for splitting it up into multiple ResourceSlice objects if there are a lot of GPUs.

Reviewer: For MIG there are 25 possible MIG devices + 1 full GPU so 26*8=208 total devices for a DGXA100 machine, for example. This is already spilling into more than one slice due to the 128 device limit. With the initial design, my plan was to create one slice per GPU and its MIG (so 8 slices total with 26 devices on each). The one downside being that I need to duplicate all of the mixins in each slice since they cannot span slices.

  //
  // +optional
  // listType=atomic

>   SharedCapacityPools []CapacityPool

Author: I'm not really sure if the `Shared` prefix here adds any value. I think maybe we should just remove it.

Reviewer 5: Yeah, it's iffy. But I want to differentiate between this and [per-device allocatable resources](https://github.com/kubernetes/enhancements/pull/5104#discussion_r1943844212), which are for user consumption.

Author: I did remove the `Shared` prefix in the updated design. I think it works pretty well without it, but let me know if you disagree.

  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting capacities from those provided via other mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Capacity map[QualifiedName]DeviceCapacity

  // SharedCapacityConsumed defines a list of references to
  // shared capacities defined in the same ResourceSlice. Each
  // of the references points to a named shared capacity and
  // includes the quantity of that capacity that the devices
  // using this mixin will need.
  //
  // During allocation, a device can only be allocated if
  // sufficient shared capacity is available to meet the
  // capacity needed by all the references provided.
  //
  // Any capacity references here will be overridden by references
  // defined in the device.
  //
  // The maximum number of consumed capacities is 32.
  //
  // +optional

>   SharedCapacityConsumed []MixinCapacityConsumedRef 

Reviewer: I don't like the idea of a mixin being able to define capacity comsumption. I explicitly disallowed this in my design because it hides too many of the details when you are looking at the device definition itself (or now the pool definition). Mixins are useful for declaring the set of capacities that should be made available by a pool, but it should be the pool (or device) itself that defines what (other) pool a given capacity is pulled from.

Author: So I think mixins in general hides some detail, since understanding the device/pool requires looking at other objects. And I think the more of these constructs we add, the more cognitive effort is required to understand what is going on. The tradeoff is obviously that fewer of these constructs means more duplication.

I'm not sure if I agree that capacity consumption on its own is worse in mixins than attributes and capacity. The way this design allows mixins to define the consumed capacity without specifying the pool makes things a little more complicated, but on the other hand I think the way it allows users to define profiles (to use the NVIDIA terminology) for partitions that can be reused across pools can also make it easier to understand if used wisely. But misuse of the mixin constructs from this design certainly makes it possible to create structures that will be hard to understand.

Reviewer: Thinking about it more, I'm not even sure what it means for a mixin to include consumed capacity. By definition, all capacity must ultimately be pulled from some concrete device (or pool in the new design). So are you saying that mixins can actually reference a concrete pool, that itself may include some other set of mixins? Can you provide a simple example here in the comments where this becomes particularly useful (and without it it becomes overly verbose)?

  // ResourceSlice.
  //
  // The maximum number of mixins that can be included is 8.
  //
  // +optional
  // +listType=atomic
  Includes []DeviceMixinRef

  // Attributes defines the set of attributes for this device.
  // The name of each attribute must be unique in that set.
  //
  // To ensure this uniqueness, attributes defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting attributes from those provided via mixins are
  // overwritten by the ones provided here.
  //
  // The maximum number of attributes and capacities combined is 32.
  //
  // +optional
  Attributes map[QualifiedName]DeviceAttribute

  // Capacity defines the set of capacities for this mixin.
  // The name of each capacity must be unique in that set.
  //
  // To ensure this uniqueness, capacities defined by the vendor
  // must be listed without the driver name as domain prefix in
  // their name. All others must be listed with their domain prefix.
  //
  // Conflicting capacities from those provided via other mixins are
  // overwritten by the ones provided here.
  //

>   // The maximum number of attributes and capacities combined is 32.

Reviewer 5: I do worry 32 is too low. How many are on, say, a MIG partition today?

Author: So for the top-level partition (gpu-0) in the NVIDIA A100 example there are 12 attributes and 13 capacities. Note that this includes capacities like `memorySlice0...memorySlice7`. So a total of 25, which is kinda close to the max number, but also maybe not quite what it would look like for a real device.

This number is not changes in this PR, so maybe @klueska has some input here.

Reviewer 5: @klueska had an "intrange" resource type concept in an early prototype. So rather than listing out memorySlice0..7, you would have a single intRange capacity [0..7], and your consumers would look like [0], [1], [0..1], etc.

This would be super useful to compact the inferentia example, too. And maybe even the TPU one if we could do, [0,1], [0,2] (ie, tuples not just ranges).

But parsing that as a resource.Quantity would be a major change since that's used all over.

Author: Ah, I was thinking that exposing the actual memory slices as capacities might not really be necessary? But I might be wrong here. What is the use-case for that?

Reviewer 5: I don't think they need to be exposed to the user as capacities, but they are needed for the shared consumption for sure. I do know though that external controllers may want to "pin down" to a particular partition, if they don't want to let the scheduler pick the partition. I think other attributes probably service that purpose (for example, partition name), but I can imagine someone using the memorySliceX in a CEL expression for that purpose.

  //
  // +optional

>   Capacity map[QualifiedName]DeviceCapacity

Reviewer: As I mentiond in another comment. I don't think devices should have the ability to define their own intrinsic capacities. *All* capacity of a device should be pulled from (one or more) pools.

In my mental model with this new pool design, there would be AttributeMixins and CapacityMixins. Devices can  include AttributeMixins and CapacityMixins, but ResourcePools can *only* include CapacityMixins.

When constructing a device you then include its set of desired attribute mixins, and then (in a separate field) list the set of CapacityMixins you want attributed to the device -- each of which *must* including a reference to a pool where the capacities listed in that mixin come from.

Reviewer 5: I think that the internal capacity consumed by partitions being the same as the values advertised to consumers is just a coincidence of the MIG case. I don't think that will be true in the general case and I don't want to lock us into it. For example, I can imagine a device that needs some overhead associated with the internal allocation before making the resource available to consumers. Or, even in the MIG case, telling the external consumers about "memory slices" seems odd.

Maybe there's a way to build the API to simplify the MIG case but still allow the general case.

Reviewer 5: Or, someone may build a device that advertises logical resources but consumes physical ones: I can service 1000 requests per second if I consume 1 core, or 2000 requests per second if I use 2 cores. So, internally they are consuming "cores" externally they are advertising "requests per second".

Author: So with the discussion offline, I think we are converging on something like:

* DeviceMixin will include both attributes and capacity. Capacity means the advertised resource availability of a device. These can only be referenced from devices and they are referenced by providing only the name.
* CapacityPoolMixin will include only capacity. These can only be referenced from CapacityPools
* SharedCapacityConsumedMixin (we should find a better name) includes the shared capacity consumed, but does not specify a capacity pool. They can only be referenced from devices. Both the name and the capacity pool will need to be provided when they are included.

I will try to get the PR updated tonight.

Author: PR has been updated.

  //
  // +required
  Name string

  // CapacityPool defines the capacity pool that will be used
  // when resolving entries in the SharedCapacityConsumed list
  // in the mixin.
  //
  // The value is only needed if the referenced mixin has
  // at least one entry in the SharedCapacityConsumed list.
  //
  // +optional
  CapacityPool string
}

// CapacityPoolMixinRef defines a reference from a capacity pool
// to a capacity pool mixin.
type CapacityPoolMixinRef struct {
  // Name is the name of a CapacityPoolMixin.
  //
  // +required
  Name string
}

// MixinCapacityConsumedRef defines a reference from a mixin
// to a capacity in an unspecified pool. This allows the same
// mixin to be used across multiple pools since the pool is
// defined when the mixin is referenced from the device.
type MixinCapacityConsumedRef struct {
  // Name defines the name of the shared capacity that
  // is being referenced.
  //
  // Name must be unique for the references from a single
  // device, but multiple devices can reference the same
  // shared capacity.
  //
  // +required
  Name string

  // Capacity defines the capacity that is required by
  // the device.
  //
  // +required

>   Capacity resource.Quantity

Reviewer: Should this be a DeviceCapacity (to match what is held in an actual CapacityPool) and not just a resource.Quantity?

Author: Yeah, I missed this.

}

// CapacityConsumedRef defines a reference from a device to
// a shared capacity in a capacity pool.
type CapacityConsumedRef struct {
  // CapacityPool defines the shared pool in which the capacity
  // referenced exists. It must point to an existing
  // capacity pool in the same ResourceSlice.
  //
  // +required
  CapacityPool string
  
  // Name defines the name of the shared capacity that
  // is being referenced.
  //
  // Name must be unique for the references from a single
  // device, but multiple devices can reference the same
  // shared capacity.
  //
  // +required
  Name string

  // Capacity defines the capacity that is required by
  // the device.
  //
  // +required

>   Capacity resource.Quantity

Reviewer: Should this be a DeviceCapacity (to match what is held in an actual CapacityPool) and not just a resource.Quantity?

}
```

As mentioned previously, the main features being added here are (1) the ability
to include a set of mixins in a device or capacity pool definition, and (2) the
ability to express that multiple devices draw from the same pool of capacity, so

> allocation of one device might make other device unallocatable.

Reviewer 5: s/other device/other devices/

The implementation of these concepts requires several changes to the
ResourceSlice API:

1. It introduces three new fields on the `ResourceSliceSpec`. First, `CapacityPools`
   defines a list of `CapacityPools`, each of which defines a set of capacities
   that devices can draw from. `CapacityPoolMixins`, `DeviceMixins`, and

>    `DeviceCapacityConsumptionMixins` defines mixins for capacity pools and devices.

Reviewer: ```suggestion
   `DeviceCapacityConsumptionMixins` define mixins for capacity pools and devices.
```

> 1. The `SharedCapacityPools` field is a list of named `CapacityPool`s. Each

Reviewer: ```suggestion
1. The `CapacityPools` field is a list of named `CapacityPool`s. Each
```

      cudaComputeCapability:
        version: 8.0.0
- name: common-gpu-capacities
  capacity:
    memory: 40Gi
  sharedCapacityConsumed:
  - name: memory

>     capacity: 40Gi

Reviewer: I'm not sure I understand this:
```
- name: common-gpu-capacities
  capacity:
    memory: 40Gi
  sharedCapacityConsumed:
  - name: memory
    capacity: 40Gi
```

What is the distinction between the `capacity` of 40GB of memory and the `sharedCapacityConsumed` of 40GB of memory? Is this to support the case where some capacity might be local to the mixin, but other capacity comes from elsewhere? What does it even mean for capacity to be local to a mixin?

Author: The idea is to separate the capacity that is consumed from the shared pool from what is advertised as capacity available to workloads. So in these scenarios, we would expect the `sharedCapacityConsumed` to be larger than the `capacity`. And they obviously don't necessarily have to have matching names.

devices:
- name: gpu-0
  composite:
    includes:
    - name: system-attributes
    - name: common-gpu-attributes
    - name: common-gpu-capacities

>       capacityPool: gpu-0-pool

Reviewer: Just having `capacityPool` tacked on here like this feels unreadable to me. Is this saying that any capacities listed in the `sharedCapacityConsumed ` section of the `common-gpu-capacities` mixin should be pulled from `gpu-0-pool`?

Author: Correct. I do agree that it reduces readability somewhat, but it makes it possible to reuse device mixins across different devices. I think the example in ["Extending device and capacity pool with a set of mixins"](https://github.com/mortent/enhancements/tree/ClarifyPartitionableDevices/keps/sig-scheduling/4815-dra-partitionable-devices#extending-device-and-capacity-pool-with-a-set-of-mixins) is good example of where it is useful, as we can use the same `common-gpu-capacities` mixin for all the different devices, even though they belong to different pools.

I think this is related to your [previous comment](https://github.com/kubernetes/enhancements/pull/5066#discussion_r1944615581) in that I think we either should allow `sharedCapacityConsumed` in device mixins with the ability to provide the capacity pool when it is referenced from the device (so what is in the current proposal), or we don't allow `sharedCapacityConsumed` in mixins at all. I think it is less useful to have them in a mixin if that also ties the mixin to a specific capacity pool.

        int: 0
      uuid:
        string: GPU-4cf8db2d-06c0-7d70-1a51-e59b25b2c16c

>     consumesCapacity:

Reviewer 2: Not really about this line...

Each of the GPUs is shown with attributes "index" and "minor". I think having just one per-device attribute (the "uuid") is sufficient and easier to understand.

As it stands now, as I reader I am trying to understand what "index" and "minor" mean although I shouldn't bother yet.

Reviewer: yes, for the example, no need to overcomplicate it I guess...

        string: GPU-662077db-fa3f-0d8f-9502-21ab0ef058a2
```

As you can see, a capacity pool mixin is created and it defines the
capacities available for each gpu. This is then referenced in four
separate shared capacity pools, each of which maps to a single physical
GPU.
Then, three mixins are created called "system-attributes",
"common-gpu-attributes", and "common-gpu-capacities", which all get included in
the definitons of the actual GPU devices themselves, along with their own
device-specific attributes (and device-specific capacities if there were any). Note
that when the "common-gpu-capacities" mixin are referenced in each device,
the name of the capacity pool is also provided.

With this in place, the scheduler can parse these device definitions and
"flatten" them into a set devices that pull capacity from a set of pools. The

> scheduler can use this information to allocate partitionable devices without

Reviewer 5: s/partitionable devices/device partitions/

capacity. It also sets up a device mixin for the partitions that defines the
available capacity of the device (available to users) and the consumed capacity
(how much it draws from the capacity pool). This allows the definition of each
partition to be quite compact.

After all mixins have been "flattened", the scheduler end up with a list
of devices that reference a set of pools. It must track the available capacity
in all capacity pools as devices gets allocated and deallocated, making sure that
the committed capacity never exceeds what is available.

Allocated resources are tracked in the ResourceClaims and that is where the
scheduler currently gets information about which devices are available. Using
this information, the scheduler can compute the available capacity for each
resource pool. Since each device specifically defines the capacity it needs,
there is no ambiguity about what capacity is being used by an allocated device.

When the scheduler tries to allocate devices for a new ResourceClaim, it will
go through the list of devices and check which of them are available. Since it
knows the free capacity in all capacity pools, it can easily check whether a

> specific device can be allocated.

Reviewer 5: Add something like:

Note that since the scheduler will go through these in order and select the *first* (not *best*) fit, it is recommended that `ResourceSlice`s include devices in smallest-to-largest order. Otherwise, the largest one will always be allocated. This may be improved through scoring mechanisms, but that is outside of the scope of this KEP.

And maybe link to the scoring issues.

Reviewer 5: (As an aside, I suppose in theory we could add something in the claim that let the user specify the sort order to use here, rather than just taking what is in the slice. But that sounds...hard and confusing.)

        int: 3
      uuid:
        string: GPU-662077db-fa3f-0d8f-9502-21ab0ef058a2
    consumesCapacity:
    - capacityPool: gpu-3-pool
      includes:
      - name: common-gpu-consumption
```

As you can see, a capacity pool mixin is created and it defines the
capacities available for each gpu. This is then referenced in four
separate capacity pools, each of which maps to a single physical
GPU.
Then, three device mixins are created called "system-attributes",
"common-gpu-attributes", and "common-gpu-capacities", which all get included in
the definitons of the actual GPU devices themselves, along with their own
device-specific attributes (and device-specific capacities if there were any). 
A device capacity consumption mixin is also created called "common-gpu-consumption"
which is referenced from all the devices together with a reference to the
capacity pool from which the capacities will be drawn.

With this in place, the scheduler can parse these device definitions and

> "flatten" them into a set devices that pull capacity from a set of pools. The

Reviewer 3: Can we include something like a kubectl cmd (not sure where it should go) or even a non-default plugin which does the flatten operation and prints the flattened thing?

Maybe we actually want a generic verb for this which fetches related things and prints them all together?  `kubectl flatten gateway foo`  

@kubernetes/sig-cli-leads

Reviewer 2: We have a similar concern over in https://github.com/kubernetes/enhancements/pull/5034: I definitely see a need to improve the CLI to help users make sense of the information that is available because it's not all in one object or even a single type. But not for alpha, and perhaps even as a separate KEP if it's considered non-blocking for graduation of the functionality itself.

scheduler can use this information to allocate device partitions without
overcommitting the capacity of the physical devices.

### Defining device partitions in terms of consumed capacity in a composite device

A simple example demonstrating how capacity pools can be used
to define multiple, allocatable partitions of a single overarching device can be
seen below.

```yaml
capacityPools:
- name: gpu-0-pool
  capacity:
    memory: 
      value: 40Gi
deviceMixins:
- name: gpu-partition-capacity
  capacity:
    memory: 
      value: 10Gi

> deviceCapacityConsuptionMixins:

Reviewer 2: ```suggestion
deviceCapacityConsumptionMixins:
```

"profile". These resources include things like "memory", "number of JPEG
engines", "discrete memory slices on the physical GPU die", etc.

Using the capacity pool to represent the "bag" of capacities, we can define
a set of devices that pull specific capacity from that pool.

> For example, capacity pool for an NVIDIA A100 GPU looks as follows:

Reviewer 5: s/, capacity/, a capacity/

a set of devices that pull specific capacity from that pool.

For example, a capacity pool for an NVIDIA A100 GPU looks as follows:
```yaml
sharedCapacityPools:
- name: gpu-0-pool
  capacity:
    copy-engines: "7"
    decoders: "5"
    encoders: "0"
    jpeg-engines: "1"
    memory: 40Gi
    memorySlice0: "1"
    memorySlice1: "1"
    memorySlice2: "1"
    memorySlice3: "1"
    memorySlice4: "1"
    memorySlice5: "1"
    memorySlice6: "1"
    memorySlice7: "1"
    multiprocessors: "98"
    ofa-engines: "1"
```

Three example devices representing MIG partitions can be defined as follows:
```yaml
sharedCapacityPools:
- name: gpu-0-pool
  ...
deviceMixins:
- name: mig-1g.5gb
  capacity:
    copy-engines: "1"
    decoders: "0"
    encoders: "0"
    jpeg-engines: "0"
    memory: 4864Mi
    memorySlice0: "1"
    multiprocessors: "14"
    ofa-engines: "0"
  sharedCapacityConsumed:
  - name: copy-engines
    capacity: "1"
  - name: decoders
    capacity: "0"
  - name: encoders
    capacity: "0"
  - name: jpeg-engines
    capacity: "0"
  - name: memory
    capacity: 4864Mi
  - name: memorySlice0
    capacity: "1"
  - name: multiprocessors
    capacity: "14"
  - name: ofa-engines

>     capacity: "0"

Reviewer: I still can't wrap my head around what it means for a mixin to consume shared capacity from somewhere. Is it not enough to simply list the capacities in the mixing and then when you include a reference to a pool at the point you include the mixin in a deivce, that indicates which pool that capacity is pulled from.

Author: Closing this since we have updated this part of the design.

available in the full GPU across its "memorySlice0" and "memorySlice1"
dimensions respectively.

Note that the use of the two device mixins that reprents the two MIG profiles
allows us to define the "shape" of the devices once and use them to define
the actual devices in just a few lines of yaml.
Also, since the device mixin representing the profiles doesn't have a
direct reference to a specific capacity pool, they can be re-used if we wanted
to represent a second NVIDIA A100.

A comprehensive example of the actual MIG partitions that would be created for
a 2 GPU DGXA100 server that ties together the concepts of both mixins and
`ConsumesCapacityFrom` can be seen below.

```yaml

> ******* WILL BE UPDATED ONCE WE MORE FEEDBACK ON THE DESIGN *******

Reviewer 2: Ready to be updated?
